web/packages/teleport/src/AppLauncher/AppLauncher.tsx

```diff
@@ -104,7 +138,10 @@ export function AppLauncher() {
       const session = await service.createAppSession(params);
 
       // Set all the fields expected by server to validate request.
-      const url = getXTeleportAuthUrl({ fqdn, port });
+      const url = getXTeleportAuthUrl({
+        publicAddr: publicAddr,
+        port,
+      });
```

Comment:
I don't understand this change, looks like you are switching the redirect url from the fqdn to the app public_addr, which seems wrong to me. Let me write out an example and let's see if we agree on what we think happens and what we think *should* happen.

Customer has two clusters, `root.example.com` and `leaf.example.com`.
There's a wildcard DNS record for *.root.example.com that points to the root proxy.
There's a wildcard DNS entry for *.leaf.example.com that points to the leaf proxy.
The root proxy has a wildcard TLS cert for *.root.example.com
The leaf proxy has a wildcard TLS cert for *.leaf.example.com

In the leaf cluster there is an app with name `dashboard` and `public_addr: dashboard.leaf.example.com`

User `alice` is logged in to the web UI on cluster `root.example.com`. In the root cluster web UI they navigate to the resources page for `leaf.example.com`, and click to launch the app `dashboard`

Before this change, IIUC, we would redirect to `dashboard.root.example.com`. DNS resolves this to the root proxy which will handle the request, the user is already logged in, as long as the root proxy actually handles the request correctly it should all work fine. I'm assuming the root proxy is supposed to be able to handle this

After this change it looks like we will redirect to the leaf app's public_addr `dashboard.leaf.example.com`. This resolves to the leaf proxy, where the user is not logged in and may not be able to log in. There is probably no app session on this cluster and it probably doesn't work

I think either:
1. I'm wrong about what this is doing, or
2. I'm right, this case wasn't considered yet and this is probably broken, or
3. We want to call it a misconfiguration to set the leaf app public_addr to an address that doesn't resolve to the root proxy

Does this make sense and can you help me figure this out?

Reply:
I suggested the change.
>  There is probably no app session on this cluster and it probably doesn't work


This is correct that the session doesn't exist. If an app session doesn't exist, then the app auth exchange takes care of this and creates the session (which is what seems like its trying to do, but missing the state cookie due to mismatched domain). Admittedly, I might be missing context on the desired behavior here.   

In the dashboard scenario you described above, what happens if both root and leaf have a dashboard.CLUSTER_HERE.example.com if we always resolve to the root? I'm unsure what the desired outcome should be.

Reply:
Maybe/probably the user can't even log in to the leaf cluster though

I assumed we had the whole utils.AssembleAppFQDN thing so that the root web UI can always redirect leaf apps to itself and explicitly ignore the leaf app public_addr, and that the root proxy would handle forwarding the request along correctly

Reply:
> In the dashboard scenario you described above, what happens if both root and leaf have a dashboard.CLUSTER_HERE.example.com if we always resolve to the root? I'm unsure what the desired outcome should be.

Obviously this is an edge case we need to handle, I have been assuming it connects to the app you last clicked on based on the cookie/state/token thing (im clearly not a web dev) set on that URL

Reply:
Hmm, fair enough. I think the mismatch of cookie domains during authentication is the real crux here in that case (which hasn't seemed to change for awhile but might have been broke for awhile too?) 

@capnspacehook can you double check the process in the app auth exchange? 

The RFD might help too (I'm unsure if leafs were included) https://github.com/gravitational/teleport/blob/87a0a1c26acf78a1630e1ba63b3f2e591ff559b4/rfd/0103-application-access-web-ui-auth-flow.md

Reply:
IMO this current change is directionally the opposite of what we want. We should always set cookies for the FQDN and always redirect to the FQDN

Reply:
Agreed. Just for additional context in the comments, the app auth exchange (using your example) is currently starting on dashboard.root and trying to finish on dashboard.leaf and that causes the state cookie to be missing.

Reply:
> Obviously this is an edge case we need to handle, I have been assuming it connects to the app you last clicked on based on the cookie/state/token thing (im clearly not a web dev) set on that URL

I'm fairly sure that was the case when I was testing Grzegorz's implementation of app access in Connect a couple of months ago.

Reply:
It was the case that it was handled correctly or incorrectly?

Reply:
i think it's hard to say exactly what's correct or incorrect, but it sounds like @ravicious was saying it used to connect to the last app you clicked on between two apps with the same name in root/leaf. I vaguely remember observing this too and being a bit surprised it actually "worked" like that

we should probably double check what this did around 16.0 and try to follow that, unless it was already broken that far back, but i do remember things working when i was testing for VNet

Reply:
I'll double check. It would have broke in 16.2 so 16.0 isn't a bad spot to look at. I'd probably the extra resolve option I added that might be resolving things incorrectly (or just "early" via root first instead of by some param). My guess it is worked as you expected before the regression otherwise we would have heard about it sooner. 

Thanks for all the input everyone!

Reply:
I was incorrect, my automatic redirects went in 16.4, not 16.2. 
regardless, I did check download 16.1.8 (teleport and Connect) and the root vs. leaf name issue is also present there. if i have `client.cluster.root` and `client.cluster.leaf` and try to open `client.cluster.leaf`, i eventually get directed to the `client.cluster.root` app instead. 

I confirmed the behavior but haven't bisected yet to figure out the cause.

Reply:
I'll bisect that and report back

Reply:
I did a lot of digging into this, and it seems this behavior has been present since at least v14.0.0. So I suppose it likely never worked correctly in this use case (root and leaf both have an app of the same name). Note that on this branch the behavior works as intended in that the leaf app is resolved correctly when it is clicked on from the web UI in the root cluster. I didn't have to login to the leaf cluster at all for this to work, so could this be actually the behavior we want? @nklaassen

Reply:
> I did a lot of digging into this, and it seems this behavior has been present since at least v14.0.0

What behaviour? Do you mean that when a root and leaf app have the same name you connect to the root app? I'm less concerned about this case, it seems uncommon, as long as we're not making a breaking change it's okay

> Note that on this branch the behavior works as intended in that the leaf app is resolved correctly when it is clicked on from the web UI in the root cluster

What exactly do you mean by resolved "correctly"?

Let me try this out and report back

Reply:
> What behaviour? Do you mean that when a root and leaf app have the same name you connect to the root app?

Yes, as far as I can tell this has never worked.

> What exactly do you mean by resolved "correctly"?

If a apps exist on both root and leaf clusters with the same name, clicking on the leaf app while on logged into the root cluster will redirect you to the leaf app. Previously you would be redirected to the root app in this case.

Reply:
Okay so I did a test on 16.0.0-alpha.2 which is an old version I happened to have saved, and on your branch. Everything I tried worked how I would expect on 16.0 and things on this branch were broken.

I have clusters `one.private:3080` and `two.private:5080`, one.private is the root, two.private is the leaf.

I have an app named `twotest` in `two.private`.
- On 16, if I click on this from the root web UI, I get directed to `twotest.one.private:3080` and it works perfectly.
- If I try the same on your branch, I get directed to `twotest.two.private:3080`, which is the wrong proxy address (I can't log in to the leaf) and it's the root cluster port. Doesn't work.

I also have an app named `netcat` in both clusters.
- On 16, if I click the leaf cluster version of this app from the root web UI, I get directed to `netcat.one.private:3080` and the connection actually goes to the app for the leaf cluster. Works perfectly.
- on your branch, if I click the leaf cluster version of this app from the root web UI, I get directed to `netcat.two.private:3080` which again is the wrong proxy and the port for the root cluster proxy. Doesn't work.

Reply:
On 16 where things work, clicking the leaf version of `netcat` causes my browser to fetch these in order:
- https://one.private:3080/web/launch/netcat.one.private/two-prod-2/netcat.two.private
- https://netcat.one.private:3080/x-teleport-auth?cluster=two-prod-2&addr=netcat.two.private
- https://one.private:3080/web/launch/netcat.one.private/two-prod-2/netcat.two.private?path=&state=b951f0996db4508cc41aed4a483febcb_4dd1ece10463fd4c66ea715e4d3d716d
- https://netcat.one.private:3080/x-teleport-auth?state=b951f0996db4508cc41aed4a483febcb_4dd1ece10463fd4c66ea715e4d3d716d&subject=a7fe3de4e8f5d23744511278701e2044c8804574d24dfe051f6b19ccaf40f445
- https://netcat.one.private:3080/

the cluster_name for my leaf cluster is two-prod-2. Notice all of these are subdomains of my root proxy `one.private:3080` which is correct

Reply:
I agree with your findings nic. I don't think the suggested change has been applied yet. The suggested change fixes the issues you speak of, just not the same name thing. 

https://github.com/gravitational/teleport/pull/47778#discussion_r1821160341

Reply:
thanks @avatus I just tried your patch on master (without this PR)

i did not try dependent/recursive apps, but with my root/leaf apps, everything was working as i expect

When opening app named `netcat` in either cluster, I was always directed to the root proxy `netcat.one.private`. If i clicked it from the leaf cluster page, I got connected to the leaf app. If i clicked it from the root cluster page, i got connected to the root app. If i kept the tabs open, all of them would start being directed to whichever app I clicked on most recently (which is the original behaviour)

@capnspacehook can we just switch this PR to be Michael's patch?
```diff
diff --git a/lib/web/app/handler.go b/lib/web/app/handler.go
index 38de429b9c..e81f686042 100644
--- a/lib/web/app/handler.go
+++ b/lib/web/app/handler.go
@@ -618,10 +618,14 @@ const (
 // The URL's are formed this way to help isolate the path params reserved for the app
 // launchers route, where order and existence of previous params matter for this route.
 func makeAppRedirectURL(r *http.Request, proxyPublicAddr, hostname string, req launcherURLParams) string {
+	host := hostname
+	if req.requiresAppRedirect {
+		host = req.publicAddr
+	}
 	u := url.URL{
 		Scheme: "https",
 		Host:   proxyPublicAddr,
-		Path:   fmt.Sprintf("/web/launch/%s", hostname),
+		Path:   fmt.Sprintf("/web/launch/%s", host),
 	}
 
 	// Presence of a stateToken means we are beginning an app auth exchange.
@@ -634,7 +638,7 @@ func makeAppRedirectURL(r *http.Request, proxyPublicAddr, hostname string, req l
 		v.Add("required-apps", req.requiredAppFQDNs)
 		u.RawQuery = v.Encode()
 
-		urlPath := []string{"web", "launch", hostname}
+		urlPath := []string{"web", "launch", host}
 
 		// The order and existence of previous params matter.
 		//
diff --git a/lib/web/app/middleware.go b/lib/web/app/middleware.go
index 79e1586ff2..67b189de3a 100644
--- a/lib/web/app/middleware.go
+++ b/lib/web/app/middleware.go
@@ -81,12 +81,8 @@ func (h *Handler) redirectToLauncher(w http.ResponseWriter, r *http.Request, p l
 			"https://goteleport.com/docs/application-access/guides/connecting-apps/#start-authproxy-service.")
 		return trace.BadParameter("public address of the proxy is not set")
 	}
-	host := p.publicAddr
-	if host == "" {
-		host = r.Host
-	}
 
-	addr, err := utils.ParseAddr(host)
+	addr, err := utils.ParseAddr(r.Host)
 	if err != nil {
 		return trace.Wrap(err)
 	}
```

Reply:
I tested this among all cases I can think of except automatic app redirects as I'm honestly not sure how to properly test that. Everything seems to be working, and I think the duplicate root/leaf app name issue isn't present with this change as well. I created an app with the same name on both the root and leaf clusters, but set the URIs so only the leaf app pointed to a listening web server. When I tried to navigate to the root app, I get an error as expected. When I navigate to the leaf app from the root cluster, I am redirected to the app successfully. The URL I'm redirected to is a root cluster URL, so that's what had me confused earlier at least.